### DISCUSSION TITLE
guides/streamlit-live-inference/

### DISCUSSION THREAD

**Discussion by giscus** [ultralytics/ultralytics#14911] [Docs]
# guides/streamlit-live-inference/

Learn how to set up a real-time object detection application using Streamlit and Ultralytics YOLOv8. Follow this step-by-step guide to implement webcam-based object detection.

### Code

- **Python**

```python
from ultralytics import solutions

solutions.inference()

### Make sure to run the file using the command `streamlit run <file-name.py>`
```

- **CLI**

```bash
yolo streamlit-predict
```

### Documentation

https://docs.ultralytics.com/guides/streamlit-live-inference/

**Comment by ghost** [comment]
AttributeError: module 'ultralytics.solutions' has no attribute 'inference'
Traceback:
File "C:\Users\DELL\AppData\Local\Programs\Python\Python312\Lib\site-packages\streamlit\runtime\scriptrunner\exec_code.py", line 75, in exec_func_with_error_handling
    result = func()
             ^^^^^^
File "C:\Users\DELL\AppData\Local\Programs\Python\Python312\Lib\site-packages\streamlit\runtime\scriptrunner\script_runner.py", line 574, in code_to_exec
    exec(code, module.__dict__)
File "D:\YOLO\infer.py", line 3, in <module>
    solutions.inference()
    ^^^^^^^^^^^^^^^^^^^
This is the error im getting while running this and i have installed necessary packages.

**Reply by pderrenger** [reply]
@Charan2128 it looks like you're encountering an `AttributeError` because the `inference` function is not found in the `ultralytics.solutions` module. Please ensure you are using the latest version of the Ultralytics package by running `pip install --upgrade ultralytics`. If the issue persists, try using the `yolo streamlit-predict` command directly from the CLI, which should launch the Streamlit application for live inference. For more details, you can refer to our [Streamlit Live Inference Guide](https://docs.ultralytics.com/guides/streamlit-live-inference/).

**Comment by ghost** [comment]
Thank you for your quick and kind response and the problem is sorted. But there is another issue while switching model to different task like pose,seg or cls and getting error of open cv not having func()

**Reply by glenn-jocher** [reply]
@Charan2128 i'm glad to hear the initial issue is resolved. Regarding the error with OpenCV when switching tasks (pose, seg, cls), please ensure you have the latest versions of both the Ultralytics package and OpenCV. You can update them using:

```bash
pip install --upgrade ultralytics opencv-python
```

If the issue persists, please provide more details about the error message and the specific function that is causing the problem. This will help us diagnose and address the issue more effectively. For further guidance, you can also refer to our [Streamlit Live Inference Guide](https://docs.ultralytics.com/guides/streamlit-live-inference/).

**Reply by ghost** [reply]
That was a Quick Replay and Quick fix ,Thanks for your kind support on my issues. The problem is fixed now.

**Reply by glenn-jocher** [reply]
@Charan2128 i'm glad to hear the initial issue is resolved. For the new issue with switching tasks and encountering OpenCV errors, please ensure you have the latest versions of both the Ultralytics package and OpenCV installed. You can update them using `pip install --upgrade ultralytics opencv-python`. If the problem persists, please provide the specific error message and function causing the issue so we can assist you further. For more detailed guidance, you can refer to our [Streamlit Live Inference Guide](https://docs.ultralytics.com/guides/streamlit-live-inference/).

**Comment by AdadAlShabab** [comment]
Can you please provide the stramlit .py file that we need to run?

**Reply by glenn-jocher** [reply]
@AdadAlShabab you can use the following code to run the Streamlit application for live inference:

```python
from ultralytics import solutions

solutions.inference()
```

Run the file with the command `streamlit run <file-name.py>`. For more details, please refer to the [Streamlit Live Inference Guide](https://docs.ultralytics.com/guides/streamlit-live-inference/).

**Comment by Cheikh-Ahmedand** [comment]
hey!

 Once I run the app my Laptop camera turns on,  instead of My webcam. How to use my external webcam instead of my LAPTOP camera?

Best Regards!

**Reply by glenn-jocher** [reply]
@Cheikh-Ahmedand to use your external webcam, modify the Streamlit application to specify the device index of your webcam. Typically, the built-in camera is `0`, so try using `1` or another index for your external webcam. Adjust the code like this:

```python
solutions.inference(source=1)
```

Ensure you run the application with `streamlit run <file-name.py>`.

**Reply by Cheikh-Ahmedand** [reply]
Hey !
Thanks lot for your response,  but the  code couldn't work with streamlit code.  I pasted the following code on my Laptop anconda PowerShell Prompt:

yolo streamlit-predict
solutions.inference(source=1)

Still,  my Webcam can't run, but my Laptop 's Camera.

Also, I tried the below code  in jupyter notebook,  but the problem still persists.

from ultralytics import solutions

solutions.inference()
solutions.inference(source=1)
### Make sure to run the file using command `streamlit run <file-name.py>`


Any help either from the Respectful Staff or others would be appreciated!!!

**Reply by glenn-jocher** [reply]
@Cheikh-Ahmedand to use your external webcam, ensure you specify the correct device index in your code. Try setting `source=1` or another index that corresponds to your external webcam. Run the Streamlit app with `streamlit run <file-name.py>`. If issues persist, please check your system's camera settings.

**Comment by sathyanarayanaRP** [comment]
how does the streamlit gets is input , is it possible to push frames or is it possible to integrate machine vision cameras that does not support RTSP protocals. i need comprehensive knowledge how does it work as a streaming device

**Reply by glenn-jocher** [reply]
@sathyanarayanaRP streamlit can receive input directly from a webcam or video file. For cameras that don't support RTSP, you can use USB or other compatible connections. The Streamlit app processes frames in real-time, allowing for seamless integration with various camera types. For more details, please refer to the Streamlit documentation.

**Comment by NNQHuong** [comment]
`AttributeError: module 'torch' has no attribute '__version__'
Traceback:
File "C:\Users\myname\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\streamlit\runtime\scriptrunner\exec_code.py", line 88, in exec_func_with_error_handling
    result = func()
             ^^^^^^
File "C:\Users\myname\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\streamlit\runtime\scriptrunner\script_runner.py", line 579, in code_to_exec
    exec(code, module.__dict__)
File "C:\WasteSegregation\app.py", line 1, in <module>
    from ultralytics import solutions
File "C:\Users\myname\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\ultralytics\__init__.py", line 11, in <module>
    from ultralytics.models import NAS, RTDETR, SAM, YOLO, FastSAM, YOLOWorld
File "C:\Users\myname\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\ultralytics\models\__init__.py", line 3, in <module>
    from .fastsam import FastSAM
File "C:\Users\myname\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\ultralytics\models\fastsam\__init__.py", line 3, in <module>
    from .model import FastSAM
File "C:\Users\myname\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\ultralytics\models\fastsam\model.py", line 5, in <module>
    from ultralytics.engine.model import Model
File "C:\Users\myname\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\ultralytics\engine\model.py", line 11, in <module>
    from ultralytics.cfg import TASK2DATA, get_cfg, get_save_dir
File "C:\Users\myname\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\ultralytics\cfg\__init__.py", line 12, in <module>
    from ultralytics.utils import (
File "C:\Users\myname\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\ultralytics\utils\__init__.py", line 51, in <module>
    TORCH_VERSION = torch.__version__
                    ^^^^^^^^^^^^^^^^^`
This is the error I encountered when using `from ultralytics import solutions` `solutions.inference(model =r'C:path\to\my\model.pt')` method (VSCode), can you please explain it to me? 
For further information, I've tried using the method for CLI and it works, but not with VSCode. I wonder if there's any method for CLI that allows me to use my custom trained model? It's YOLO11n-cls version. Thank you for reading this!

**Reply by glenn-jocher** [reply]
@NNQHuong the error you're encountering suggests a possible issue with your PyTorch installation. Please verify that PyTorch is properly installed and compatible with your Python version by running:

```python
import torch
print(torch.__version__)
```

If the error persists, ensure you're using the latest version of PyTorch and Ultralytics with:

```
pip install --upgrade torch ultralytics
```

Regarding your CLI query, you can use your custom-trained YOLO11 model by specifying the model path directly, like so:

```
yolo predict model=path/to/your/model.pt source=your_input
```

For detailed examples, check the documentation [here](https://docs.ultralytics.com/modes/predict/). Let me know if additional assistance is required!

**Reply by RizwanMunawar** [reply]
Hi @NNQHuong, thank you for sharing the error log! It appears that the issue might be related to your PyTorch version. You may need to upgrade or downgrade it to ensure everything works correctly.

However, if this still doesn't resolve the issue, please share your code here for further testing.

Thanks,
Ultralytics Team!

**Reply by NNQHuong** [reply]
Thanks for your infomative reply, it worked smoothly! I'm kind of new to this field anyway, about 6 months , I bet you can see it in my words haha, so thanks for your patience too!

**Reply by glenn-jocher** [reply]
Thank you for your kind words and for sharing your progress! The YOLO community and Ultralytics team are always here to support your learning journey. If you encounter any other issues, feel free to ask. Keep exploring and building amazing things!

**Comment by Tech-Watt** [comment]
Is it posible to use different camera index ?. It picks up the default computer webcam and there's no means of changing it.

**Reply by glenn-jocher** [reply]
To use a different camera index, you can modify the video source parameter in your code like this:

```python
inf = solutions.Inference(model="yolo11n.pt", source=1)  # Change index as needed
```

This allows you to specify alternate camera devices while maintaining Streamlit's web interface functionality. For implementation details, see the Streamlit Inference documentation.

**Comment by j3dd4h** [comment]
the manual execution using rknn model (yolo predict model='./best_rknn_model' source='ppe.jpg')  is working  but got an error when running under Streamlit.

TypeError: model='best_rknn_model/best-rk3588.rknn' is not a supported model format. Ultralytics supports: ('PyTorch', 'TorchScript', 'ONNX', 'OpenVINO', 'TensorRT', 'CoreML', 'TensorFlow SavedModel', 'TensorFlow GraphDef', 'TensorFlow Lite', 'TensorFlow Edge TPU', 'TensorFlow.js', 'PaddlePaddle', 'MNN', 'NCNN', 'IMX', 'RKNN') See https://docs.ultralytics.com/modes/predict for help.

**Reply by RizwanMunawar** [reply]
@j3dd4h Hi, thanks for sharing the details. We currently don't support the rknn model, but you can adjust this [check in code](https://github.com/ultralytics/ultralytics/blob/3a9dd219d8a5ca43a0e969bafb81f9378528f94c/ultralytics/solutions/streamlit_inference.py#L170) to make the application work for rknn models. We will add support for this soon.

**Reply by j3dd4h** [reply]
Ok, got it

**Reply by RizwanMunawar** [reply]
@j3dd4h opened the Pull request here: https://github.com/ultralytics/ultralytics/pull/22212
